libraries/AP_Compass/AP_Compass_BMM350.cpp

```diff
@@ -0,0 +1,557 @@
+#define BMM350_Z_INA_GAIN               31.0f
+#define BMM350_ADC_GAIN                 (1.0f / 1.5f)
+#define BMM350_LUT_GAIN                 0.714607238769531f
+#define BMM350_POWER                    ((float)(1000000.0 / 1048576.0))
+
+#define BMM350_XY_SCALE                 (BMM350_POWER / (BMM350_XY_SENSITIVE * BMM350_XY_INA_GAIN * BMM350_ADC_GAIN * BMM350_LUT_GAIN))
+#define BMM350_Z_SCALE                  (BMM350_POWER / (BMM350_Z_SENSITIVE * BMM350_Z_INA_GAIN * BMM350_ADC_GAIN * BMM350_LUT_GAIN))
+#define BMM350_TEMP_SCALE               (1.0 / (BMM350_TEMP_SENSITIVE * BMM350_ADC_GAIN * BMM350_LUT_GAIN * 1048576))
+
+extern const AP_HAL::HAL &hal;
+
+AP_Compass_Backend *AP_Compass_BMM350::probe(AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                              bool force_external,
+                                              enum Rotation rotation)
+{
+    if (!dev) {
+        return nullptr;
+    }
+    AP_Compass_BMM350 *sensor = new AP_Compass_BMM350(std::move(dev), force_external, rotation);
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_BMM350::AP_Compass_BMM350(AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                       bool force_external,
+                                       enum Rotation rotation)
+    : _dev(std::move(dev))
+    , _force_external(force_external)
+    , _rotation(rotation)
+{
+}
+
+int32_t AP_Compass_BMM350::fix_sign(uint32_t val, uint8_t bit)
+{
+    int32_t power = 0;
+    int32_t ret = (int32_t)val;
+
+    switch (bit) {
+        case BMM350_SIGNED_12_BIT:
+            power = 2048; // 2^11
+            break;
+        case BMM350_SIGNED_16_BIT:
+            power = 32768; // 2^15
+            break;
+        case BMM350_SIGNED_24_BIT:
+            power = 8388608; // 2^23
+            break;
+        default:
+            power = 0;
+            break;
+    }
+
+    if (ret >= power) {
+        ret -= (power * 2);
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Read out OTP(one-time programmable memory) data of sensor which is the compensation coefficients
+ * @see https://github.com/boschsensortec/BMM350-SensorAPI
+ */
+bool AP_Compass_BMM350::read_otp_data()
+{
+    uint8_t data = 0;
+    uint8_t msb = 0;
+    uint8_t lsb = 0;
+    uint16_t otp_data[BMM350_OTP_DATA_LENGTH];
+
+    for (uint8_t index = 0; index < BMM350_OTP_DATA_LENGTH; index++) {
+        // Set OTP address
+        data = (BMM350_OTP_CMD_DIR_READ | index);
+        if (!_dev->write_register(BMM350_REG_OTP_CMD, data)) {
+            return false;
+        }
+
+        // Wait OTP status be ready
+        do {
+            hal.scheduler->delay_microseconds(300);
+            // Read OTP status
+            if (!read_bytes(BMM350_REG_OTP_STATUS, &data, 1) || ((data & BMM350_OTP_STATUS_ERROR_MASK) != 0)) {
+                return false;
+            }
+        } while (!(data & BMM350_OTP_STATUS_CMD_DONE));
```

Comment:
We're relying on one of the read to fail or the error bit to be set to break out of this loop.  That's not good as the device may get into a state where it won't do either of those.

Should have some other way of getting out of this loop and returning false; perhaps a simple loop counter.  You're essentially polling for the result here, so perhaps a `bool poll_otp_register` method?